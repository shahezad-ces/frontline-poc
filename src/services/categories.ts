import {
  GET_CATEGORIES_QUERY,
  GET_CATEGORY_DETAILS_QUERY,
} from "@frontline/graphql";
import { query } from "@frontline/libs/ApolloClient";
import { Category } from "@frontline/types";

export const getCategories = async (): Promise<Category[] | undefined> => {
  try {
    const { data } = await query<{ categories: Category[] }>({
      query: GET_CATEGORIES_QUERY,
      fetchPolicy: "cache-first",
    });

    return data?.categories;
  } catch (error) {
    console.error("GraphQL Error in getCategories:", error);
    return undefined;
  }
};

export const getCategoryDetails = async (
  id: number
): Promise<Category | undefined> => {
  if (!id) return undefined;
  try {
    const { data } = await query<{ category: Category }>({
      query: GET_CATEGORY_DETAILS_QUERY,
      variables: { id },
      fetchPolicy: "cache-first",
    });

    return data?.category;
  } catch (error) {
    console.error("GraphQL Error in getCategoryDetails:", error);
    return undefined;
  }
};
