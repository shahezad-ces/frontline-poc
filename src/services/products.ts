import {
  GET_PRODUCTS_QUERY,
  GET_PRODUCT_DETAILS_QUERY,
} from "@frontline/graphql";
import { query } from "@frontline/libs/ApolloClient";
import type { Product } from "@frontline/types";

export interface GetProductsParams {
  categoryId?: number;
  price?: number;
  price_min?: number;
  price_max?: number;
  offset?: number;
  limit?: number;
}

export const getProducts = async (
  params: GetProductsParams
): Promise<Product[] | undefined> => {
  const {
    categoryId,
    price,
    price_min,
    price_max,
    offset = 0,
    limit = 20,
  } = params;
  const variables = {
    offset,
    limit,
    ...(categoryId !== undefined && { categoryId }),
    ...(price !== undefined && { price }),
    ...(price_min !== undefined && { price_min }),
    ...(price_max !== undefined && { price_max }),
  };

  const { data } = await query<{ products: Product[] }>({
    query: GET_PRODUCTS_QUERY,
    variables,
    fetchPolicy: "cache-first",
  });

  return data?.products;
};

export const getProductDetails = async (
  productId: number
): Promise<Product | undefined> => {
  try {
    const { data } = await query<{ product: Product }>({
      query: GET_PRODUCT_DETAILS_QUERY,
      variables: { id: productId },
      fetchPolicy: "cache-first",
    });

    return data?.product;
  } catch (error) {
    console.error("GraphQL Error:", error);
    return undefined; // fallback
  }
};
