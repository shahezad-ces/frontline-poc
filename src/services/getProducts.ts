import { GET_PRODUCTS_QUERY } from "@frontline/graphql/products";
import { fetchGraphQL } from "@frontline/libs/apolloClient";
import { Product } from "@frontline/types/product";

export async function getProducts(): Promise<Product[]> {
  const variables = {
    offset: 0,
    limit: 20,
  };

  const data = await fetchGraphQL<{ products: Product[] }>({
    query: GET_PRODUCTS_QUERY,
    variables,
  });

  return data.products;
}
