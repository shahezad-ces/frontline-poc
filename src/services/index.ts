import { getCategories, getCategoryDetails } from "./categories";
import { getProducts, getProductDetails } from "./products";

export { getCategories, getCategoryDetails, getProducts, getProductDetails };
