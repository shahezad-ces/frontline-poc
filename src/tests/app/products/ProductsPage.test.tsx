import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import ProductsPage from "@frontline/app/products/page";
import { getProducts, getCategoryDetails } from "@frontline/services";

// Mock services
jest.mock("@frontline/services", () => ({
  getProducts: jest.fn(),
  getCategoryDetails: jest.fn(),
}));

// Mock child components to avoid client-side hook issues
jest.mock("@frontline/components", () => ({
  Categories: () => <div data-testid="categories">Categories Component</div>,
  ProductsFilters: () => <div data-testid="filters">Filters Component</div>,
  PaginationWrapper: () => (
    <div data-testid="pagination">Pagination Component</div>
  ),
  ProductCard: ({ title }: any) => <div>{title}</div>,
}));

const mockGetProducts = getProducts as jest.Mock;
const mockGetCategoryDetails = getCategoryDetails as jest.Mock;

const mockProducts = [
  { id: 1, title: "Product 1" },
  { id: 2, title: "Product 2" },
];

describe("ProductsPage (Server Component)", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("handles categoryId provided but category details not found", async () => {
    mockGetProducts.mockResolvedValue(mockProducts);
    mockGetCategoryDetails.mockResolvedValue(null);

    const ui = await ProductsPage({ searchParams: { categoryId: "99" } });
    const { asFragment } = render(ui);

    // Category name should NOT appear
    expect(screen.queryByText(/Electronics/i)).not.toBeInTheDocument();

    // Products should render
    expect(screen.getByText(/Product 1/i)).toBeInTheDocument();
    expect(screen.getByText(/Product 2/i)).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
