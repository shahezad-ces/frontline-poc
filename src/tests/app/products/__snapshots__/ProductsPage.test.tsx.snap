// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductsPage (Server Component) handles categoryId provided but category details not found 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto px-4 my-6 sm:px-6 lg:px-8 gap-6"
  >
    <h1
      class="sm:text-2xl lg:text-3xl font-bold mb-4"
    />
    <div
      class="flex flex-col lg:flex-row gap-6"
    >
      <div
        class="w-full lg:w-1/4 space-y-4"
      >
        <div
          data-testid="categories"
        >
          Categories Component
        </div>
        <div
          data-testid="filters"
        >
          Filters Component
        </div>
      </div>
      <div
        class="w-full lg:w-3/4"
      >
        <div
          class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 xl:grid-cols-4 gap-6"
        >
          <div>
            Product 1
          </div>
          <div>
            Product 2
          </div>
        </div>
      </div>
    </div>
    <div
      class="mt-6 w-full lg:w-3/4 lg:ml-auto"
    >
      <div
        data-testid="pagination"
      >
        Pagination Component
      </div>
    </div>
  </div>
</DocumentFragment>
`;
